transport/http/transport.go

```diff
@@ -88,49 +110,63 @@ func NewTransport(opts ...TransportOption) *Transport {
 		o(&cfg)
 	}
 
+	client, transport := buildClient(&cfg)
 	return &Transport{
-		once:   intsync.Once(),
-		client: buildClient(&cfg),
-		peers:  make(map[string]*hostport.Peer),
-		tracer: cfg.tracer,
+		once:                intsync.Once(),
+		transport:           transport,
+		client:              client,
+		peers:               make(map[string]*hostport.Peer),
+		tracer:              cfg.tracer,
+		socks5ProxyHostPort: cfg.socks5ProxyHostPort,
 	}
 }
 
-func buildClient(cfg *transportConfig) *http.Client {
-	return &http.Client{
-		Transport: &http.Transport{
-			// options lifted from https://golang.org/src/net/http/transport.go
-			Proxy: http.ProxyFromEnvironment,
-			Dial: (&net.Dialer{
-				Timeout:   30 * time.Second,
-				KeepAlive: cfg.keepAlive,
-			}).Dial,
-			TLSHandshakeTimeout:   10 * time.Second,
-			ExpectContinueTimeout: 1 * time.Second,
-			MaxIdleConnsPerHost:   cfg.maxIdleConnsPerHost,
-		},
+func buildClient(cfg *transportConfig) (*http.Client, *http.Transport) {
+	transport := &http.Transport{
+		// options lifted from https://golang.org/src/net/http/transport.go
+		Proxy: http.ProxyFromEnvironment,
+		Dial: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: cfg.keepAlive,
+		}).Dial,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		MaxIdleConnsPerHost:   cfg.maxIdleConnsPerHost,
 	}
+
+	return &http.Client{Transport: transport}, transport
 }
 
 // Transport keeps track of HTTP peers and the associated HTTP client. It
 // allows using a single HTTP client to make requests to multiple YARPC
 // services and pooling the resources needed therein.
 type Transport struct {
-	lock sync.Mutex
-	once intsync.LifecycleOnce
-
-	client *http.Client
-	peers  map[string]*hostport.Peer
-
-	tracer opentracing.Tracer
+	lock                sync.Mutex
+	once                intsync.LifecycleOnce
+	transport           *http.Transport
+	client              *http.Client
+	peers               map[string]*hostport.Peer
+	tracer              opentracing.Tracer
+	socks5ProxyHostPort string
 }
 
 var _ transport.Transport = (*Transport)(nil)
 
 // Start starts the HTTP transport.
 func (a *Transport) Start() error {
 	return a.once.Start(func() error {
-		return nil // Nothing to do
+		// Explicit passing this option in will override the ProxyFromEnvironment
+		if a.socks5ProxyHostPort != "" {
+			dialer, err := proxy.SOCKS5("tcp", a.socks5ProxyHostPort, nil, proxy.Direct)
```

Comment:
instead of proxy.Direct, should we be using the semantics for our original dialer? (line 128).  It looks like the proxy.Direct is just a default net.Dialer.  This will allow us to keep the keepalive config and any configuration we add to the dialer in the future.